stmt/stmts/grant.go

```diff
@@ -0,0 +1,760 @@
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/column"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/model"
+	mysql "github.com/pingcap/tidb/mysqldef"
+	"github.com/pingcap/tidb/parser/coldef"
+	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/rset"
+	"github.com/pingcap/tidb/rset/rsets"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/db"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/stmt"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/format"
+)
+
+/************************************************************************************
+ * Grant Statement
+ * See: https://dev.mysql.com/doc/refman/5.7/en/grant.html
+ ************************************************************************************/
+var (
+	_ stmt.Statement = (*GrantStmt)(nil)
+)
+
+// GrantStmt grants privilege to user account.
+type GrantStmt struct {
+	Privs      []*coldef.PrivElem
+	ObjectType int
+	Level      *coldef.GrantLevel
+	Users      []*coldef.UserSpecification
+	Text       string
+}
+
+// Explain implements the stmt.Statement Explain interface.
+func (s *GrantStmt) Explain(ctx context.Context, w format.Formatter) {
+	w.Format("%s\n", s.Text)
+}
+
+// IsDDL implements the stmt.Statement IsDDL interface.
+func (s *GrantStmt) IsDDL() bool {
+	return true
+}
+
+// OriginText implements the stmt.Statement OriginText interface.
+func (s *GrantStmt) OriginText() string {
+	return s.Text
+}
+
+// SetText implements the stmt.Statement SetText interface.
+func (s *GrantStmt) SetText(text string) {
+	s.Text = text
+}
+
+// Exec implements the stmt.Statement Exec interface.
+func (s *GrantStmt) Exec(ctx context.Context) (rset.Recordset, error) {
+	// Grant for each user
+	for _, user := range s.Users {
+		// Check if user exists.
+		strs := strings.Split(user.User, "@")
+		userName := strs[0]
+		host := strs[1]
+		exists, err := userExists(ctx, userName, host)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if !exists {
+			return nil, errors.Errorf("Unknown user: %s", user.User)
+		}
+
+		// If there is no privilege entry in corresponding table, insert a new one.
+		// DB scope:		mysql.DB
+		// Table scope:		mysql.Tables_priv
+		// Column scope:	mysql.Columns_priv
+		switch s.Level.Level {
+		case coldef.GrantLevelDB:
+			err := s.checkAndInitDBPriv(ctx, userName, host)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		case coldef.GrantLevelTable:
+			err := s.checkAndInitTablePriv(ctx, userName, host)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		// Grant each priv to the user.
+		for _, priv := range s.Privs {
+			if len(priv.Cols) > 0 {
+				// Check column scope privilege entry.
+				err1 := s.checkAndInitColumnPriv(ctx, userName, host, priv.Cols)
+				if err1 != nil {
+					return nil, errors.Trace(err1)
+				}
+			}
+			err2 := s.grantPriv(ctx, priv, user)
```

Comment:
Any need to check grant validity first? like grant xxx, then we will not to insert rows above.